pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -54,29 +55,63 @@ type AmbientIndex struct {
 	byService map[string][]*model.WorkloadInfo
 	// byPod indexes by Pod IP address.
 	byPod map[string]*model.WorkloadInfo
+	// services are indexed by the network/clusterIP
+	services map[string]*workloadapi.Service
 
 	// Map of ServiceAccount -> IP
 	// TODO: currently, this is derived from pods. To be agnostic to the implementation,
 	// we should actually be looking at Gateway.status.addresses.
 	// This may be an external address (possibly even a DNS name we need to resolve), an arbitrary IP,
 	// or a reference to a service.
 	// If its a reference to a Service then we can find the underlying pods in that service, as an optimization.
-	waypoints map[model.WaypointScope]sets.String
+	waypoints map[model.WaypointScope]*workloadapi.GatewayAddress
 
 	// serviceVipIndex maintains an index of VIP -> Service
 	serviceVipIndex *kclient.Index[string, *v1.Service]
 }
 
 // Lookup finds a given IP address.
-func (a *AmbientIndex) Lookup(ip string) []*model.WorkloadInfo {
+func (a *AmbientIndex) Lookup(key string) []*model.AddressInfo {
 	a.mu.RLock()
 	defer a.mu.RUnlock()
 	// First look at pod...
-	if p, f := a.byPod[ip]; f {
-		return []*model.WorkloadInfo{p}
+	if p, f := a.byPod[key]; f {
+		addr := &workloadapi.Address{
+			Type: &workloadapi.Address_Workload{
+				Workload: p.Workload,
+			},
+		}
+		return []*model.AddressInfo{{Address: addr}}
 	}
+
 	// Fallback to service. Note: these IP ranges should be non-overlapping
-	return a.byService[ip]
+	// cannot distunagish between Service lookup and Workloads for a Service lookup
+	// since the same newtork/vip key is used.  Both Service and Workloads are
+	// returned in this case
+	// TODO any service to exclude in the future? for example with affinity,
+	// per-node, etc - where simply load balancing will break?
+	res := []*model.AddressInfo{}
+	for _, wl := range a.byService[key] {
+		addr := &workloadapi.Address{
+			Type: &workloadapi.Address_Workload{
+				Workload: wl.Workload,
+			},
+		}
+		res = append(res, &model.AddressInfo{Address: addr})
+	}
+	if s, exists := a.services[key]; exists {
+		addr := &workloadapi.Address{
+			Type: &workloadapi.Address_Service{
+				Service: s,
+			},
+		}
+		res = append(res, &model.AddressInfo{Address: addr})
+	} else if len(res) > 0 {
+		// there are Workloads for a Service that does not exist, which should never happen
+		log.Warnf("workload(s) found for %q but no service exists", key)
+	}
```

Comment:
i think i mislead you here sorry https://github.com/istio/istio/pull/44199/files#r1195277695